File: day-03/python/main.py
import pathlib
import string
from typing import Iterable


def common_character(strings: list[str]) -> str:
    return set.intersection(*map(set, strings)).pop()


def sum_priorities(strings: Iterable[list[str]]) -> int:
    priority_map = {
        letter: priority
        for priority, letter in enumerate(string.ascii_letters, start=1)
    }
    return sum(priority_map[common_character(group)] for group in strings)


with open(pathlib.Path(__file__).parent.parent / 'input.txt') as f:
    rucksacks = f.read().splitlines()


part_one_solution = sum_priorities(
    [rucksack[: len(rucksack) // 2], rucksack[len(rucksack) // 2 :]]
    for rucksack in rucksacks
)
part_two_solution = sum_priorities(
    rucksacks[i : i + 3] for i in range(0, len(rucksacks), 3)
)

print('Part One:', part_one_solution)
print('Part Two:', part_two_solution)

assert part_one_solution == 8018
assert part_two_solution == 2518

File: day-03/rust/src/main.rs
use std::collections::HashMap;
use std::fs;
use std::path::Path;

use itertools::Itertools;

fn common_character(strings: &[&str]) -> char {
    let common_character = &mut strings[0].to_string();
    for string in strings {
        common_character.retain(|e| string.contains(e));
    }
    common_character.chars().next().unwrap()
}

fn sum_priorities(strings: &Vec<Vec<&str>>) -> i32 {
    let mut priority_map: HashMap<char, i32> = HashMap::new();
    for (i, character) in ('a'..='z').chain('A'..='Z').into_iter().enumerate() {
        priority_map.insert(character, i as i32 + 1);
    }

    strings
        .iter()
        .map(|strings| {
            *priority_map
                .get(&common_character(strings.as_slice()))
                .unwrap()
        })
        .collect_vec()
        .iter()
        .sum()
}

fn part_one(rucksacks: &Vec<&str>) -> i32 {
    sum_priorities(
        &rucksacks
            .iter()
            .map(|rucksack| {
                let (compartment_one, compartment_two) = rucksack.split_at(rucksack.len() / 2);
                vec![compartment_one, compartment_two]
            })
            .collect_vec(),
    )
}

fn part_two(rucksacks: &Vec<&str>) -> i32 {
    sum_priorities(
        &rucksacks
            .into_iter()
            .copied()
            .chunks(3)
            .into_iter()
            .map(|rucksacks| rucksacks.collect_vec())
            .collect_vec(),
    )
}

fn main() {
    let input = fs::read_to_string(
        Path::new(env!("CARGO_MANIFEST_DIR"))
            .parent()
            .unwrap()
            .join("input.txt"),
    )
    .unwrap();
    let rucksacks = input.lines().collect();

    let part_one_solution = part_one(&rucksacks);
    let part_two_solution = part_two(&rucksacks);

    println!("Part One: {}", part_one_solution);
    println!("Part Two: {}", part_two_solution);

    assert_eq!(part_one_solution, 8018);
    assert_eq!(part_two_solution, 2518);
}
